service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/workspace/CloneCreateCloudContextFlight.java

```diff
@@ -0,0 +1,107 @@
+import java.util.Optional;
+import java.util.UUID;
+
+/** Flight for creating a cloud context for a cloned workspace */
+public class CloneCreateCloudContextFlight extends Flight {
+  public CloneCreateCloudContextFlight(FlightMap inputParameters, Object applicationContext) {
+    super(inputParameters, applicationContext);
+
+    // Flight Plan
+    // * Launch a flight to create a cloud context if necessary
+    // * Await the context flight
+    var flightBeanBag = FlightBeanBag.getFromObject(applicationContext);
+    var cloudRetryRule = RetryRules.cloud();
+    var longCloudRetryRule = RetryRules.cloudLongRunning();
+    var userRequest =
+        FlightUtils.getRequired(
+            inputParameters,
+            JobMapKeys.AUTH_USER_INFO.getKeyName(),
+            AuthenticatedUserRequest.class);
+    var sourceWorkspaceId =
+        FlightUtils.getRequired(
+            inputParameters, ControlledResourceKeys.SOURCE_WORKSPACE_ID, UUID.class);
+    var spendProfile =
+        inputParameters.get(WorkspaceFlightMapKeys.SPEND_PROFILE, SpendProfile.class);
+
+    addStep(new CreateCloudContextIdsForFutureStepsStep());
+
+    // Only create a cloud context if the source workspace has a cloud context
+    Optional<GcpCloudContext> gcpCloudContextOptional =
+        flightBeanBag.getGcpCloudContextService().getGcpCloudContext(sourceWorkspaceId);
+    if (gcpCloudContextOptional.isPresent()) {
+      SpendProfile gcpSpendProfile =
+          getSpendProfile(spendProfile, gcpCloudContextOptional.get(), flightBeanBag, userRequest);
+
+      addStep(
+          new LaunchCreateCloudContextFlightStep(
+              flightBeanBag.getWorkspaceService(),
+              CloudPlatform.GCP,
+              gcpSpendProfile,
+              ControlledResourceKeys.CREATE_GCP_CLOUD_CONTEXT_FLIGHT_ID),
+          cloudRetryRule);
+      addStep(
+          new AwaitCreateCloudContextFlightStep(
+              ControlledResourceKeys.CREATE_GCP_CLOUD_CONTEXT_FLIGHT_ID),
+          longCloudRetryRule);
+    }
+
+    Optional<AzureCloudContext> azureCloudContextOptional =
+        flightBeanBag.getAzureCloudContextService().getAzureCloudContext(sourceWorkspaceId);
+    if (azureCloudContextOptional.isPresent()) {
+      SpendProfile azureSpendProfile =
+          getSpendProfile(
+              spendProfile, azureCloudContextOptional.get(), flightBeanBag, userRequest);
+      addStep(
+          new LaunchCreateCloudContextFlightStep(
+              flightBeanBag.getWorkspaceService(),
+              CloudPlatform.AZURE,
+              azureSpendProfile,
+              ControlledResourceKeys.CREATE_AZURE_CLOUD_CONTEXT_FLIGHT_ID),
+          cloudRetryRule);
+      addStep(
+          new AwaitCreateCloudContextFlightStep(
+              ControlledResourceKeys.CREATE_AZURE_CLOUD_CONTEXT_FLIGHT_ID),
+          cloudRetryRule);
+    }
+  }
+
+  /*
+   * There is an API mismatch in the V1 APIs. If a new spend profile is passed with
```

Comment:
Another future looking note: it may be time to start making stronger assertions about having a single CC in a workspace. Technically it's possible to have more than one but in practice that is a very unsupported and likely to lead to strange bugs.

Reply:
Created https://broadworkbench.atlassian.net/browse/WOR-1626